app/src/main/java/de/westnordost/streetcomplete/quests/address/AddAddressStreetForm.kt

```diff
@@ -0,0 +1,179 @@
+    @Inject
+    internal lateinit var roadNameSuggestionsDao: RoadNameSuggestionsDao
+
+    @Inject internal lateinit var favs: LastPickedValuesStore<String>
+
+    init {
+        Injector.instance.applicationComponent.inject(this)
+        val lastPickedNames = favs.get(javaClass.simpleName)
+        defaultName = if (lastPickedNames.isEmpty()) {""} else {lastPickedNames.first}
+    }
+
+    override fun setupNameAdapter(data: List<LocalizedName>, addLanguageButton: View): AddLocalizedNameAdapter {
+
+        return AddLocalizedNameAdapter(
+                data, activity!!, getPossibleStreetsignLanguages(),
+                abbreviationsByLocale, getRoadNameSuggestions(), addLanguageButton,
+                defaultName
+        )
+    }
+
+    private fun getRoadNameSuggestions(): List<MutableMap<String, String>> {
+        return roadNameSuggestionsDao.getNames(
+                geometryToMajorPoints(elementGeometry),
+                AddAddressStreet.MAX_DIST_FOR_ROAD_NAME_SUGGESTION
+        )
+    }
+
+    private fun geometryToMajorPoints(geometry: ElementGeometry): List<LatLon> {
+        return when(geometry) {
+            is ElementPolylinesGeometry -> {
+                val polyline = geometry.polylines.first()
+                listOf(polyline.first(), polyline.last())
+            }
+            is ElementPolygonsGeometry -> {
+                // return center and one of nodes from the way constructing area
+                listOf(geometry.center, geometry.polygons.first().last())
+            }
+            is ElementPointGeometry -> {
+                listOf(geometry.center)
+            }
+        }
+
+
+    }
+
+    override fun isRejectingClose() : Boolean {
+        // if the form is complete, we will reject close, unless it is still has only the default name
+        if (isFormComplete()) {
+            if (adapter.localizedNames.size == 1 && adapter.localizedNames.first().name == defaultName) {
+                return false
+            } else {
+                return true
+            }
+        } else {
+            return false
+        }
+    }
+
+    private fun switchToPlaceName() {
+        AlertDialog.Builder(activity!!)
+                .setTitle(R.string.quest_address_street_noStreet_confirmation_title)
+                .setPositiveButton(R.string.quest_address_street_noStreet_confirmation_positive) {_, _ -> switchToPlaceNameLayout() } //
+                .setNegativeButton(R.string.quest_generic_confirmation_no, null)
+                .show()
+    }
+
+    private fun switchToPlaceNameLayout() {
+        isPlaceName = true
+        setLayout(R.layout.quest_streetname_place)
+    }
+
+    private fun setLayout(layoutResourceId: Int) {
+        val view = setContentView(layoutResourceId)
+        val buttonNameSuggestions : View = view.findViewById(R.id.nameSuggestionsButton)
+        textField = view.findViewById(R.id.name)
+
+        //TODO - use actual place names
```

Comment:
suggestions for place names could always be added later - as long as the form does not behave wrongly (i.e. show road names as suggestions for place names), that's fine.

Reply:
It now shows no suggestions when you change to PlaceName mode.